File: templates/python/http_proxy.py
"""模板:HTTP API → ACP 元件(通用代理,Python 版)。

在 ENDPOINTS 里声明「元件 id → 上游 URL/方法」即可,零业务代码。
启动:python templates/python/http_proxy.py
"""
from __future__ import annotations

import json
import os
import urllib.request
from typing import Any

from acp import AcpError, AcpErrorCode, AcpServer
from acp.component import ComponentDef

ENDPOINTS: list[dict[str, Any]] = [
    {
        "id": "http.jsonplaceholder.user",
        "description": "Fetches a user from JSONPlaceholder by id",
        "url": "https://jsonplaceholder.typicode.com/users/{id}",
        "method": "GET",
        "input_schema": {"type": "object", "properties": {"id": {"type": "integer", "minimum": 1}}, "required": ["id"]},
    },
    {
        "id": "http.jsonplaceholder.post",
        "description": "Creates a post on JSONPlaceholder",
        "url": "https://jsonplaceholder.typicode.com/posts",
        "method": "POST",
        "input_schema": {
            "type": "object",
            "properties": {"title": {"type": "string"}, "body": {"type": "string"}},
            "required": ["title"],
        },
    },
]


def _fill_url(url: str, inp: dict[str, Any]) -> str:
    for key, value in inp.items():
        url = url.replace("{" + key + "}", urllib.request.quote(str(value)))
    return url


def build_components() -> list[ComponentDef]:
    defs: list[ComponentDef] = []
    for ep in ENDPOINTS:
        def make_handle(ep: dict[str, Any]):
            def handle(inp: Any, ctx) -> Any:  # noqa: ANN001
                body = json.dumps(inp or {}).encode()
                req = urllib.request.Request(
                    _fill_url(ep["url"], inp or {}),
                    data=body if ep["method"] == "POST" else None,
                    headers={"content-type": "application/json"},
                    method=ep["method"],
                )
                try:
                    with urllib.request.urlopen(req, timeout=30) as resp:
                        return json.loads(resp.read().decode())
                except Exception as exc:  # 上游失败 → ACP 语义(spec §8:50002)
                    raise AcpError(AcpErrorCode.UPSTREAM_ERROR, f"upstream error: {exc}") from exc
            return handle

        defs.append(
            ComponentDef(
                id=ep["id"],
                name=ep["id"],
                description=ep["description"],
                input_schema=ep.get("input_schema"),
                tags=["http-proxy"],
                handle=make_handle(ep),
            )
        )
    return defs


if __name__ == "__main__":
    server = AcpServer(name="http-proxy-node", version="1.0.0")
    for comp in build_components():
        server.register(comp)
    ws_port = server.listen(port=int(os.environ.get("ACP_PORT", "8091")))
    print(f"HTTP ACP endpoint: http://localhost:{server.http_port}/acp  (WS: ws://localhost:{ws_port}/acp)")
    server.serve_forever()

File: templates/src/http-proxy.ts
/**
 * 模板:HTTP API → ACP 元件(通用代理)
 *
 * 把任意 HTTP JSON API 封装为 ACP 元件:只需在 ENDPOINTS 数组里声明
 * 「元件 id → 上游 URL/方法」,零业务代码。生产中把这里换成你的内网服务即可。
 *
 * 启动:pnpm start:http-proxy(或 node --experimental-strip-types src/http-proxy.ts)
 */
import { AcpServer, AcpError, AcpErrorCode, defineComponent } from "@xi-era/acp-sdk/server";
import type { ComponentDef } from "@xi-era/acp-sdk/server";

interface EndpointConfig {
  /** ACP 元件 id(必须匹配 ^[a-z][a-z0-9-]{0,62}(\.[a-z][a-z0-9-]{0,62}){1,3}$) */
  id: string;
  description: string;
  /** 上游 HTTP 接口 */
  url: string;
  method: "GET" | "POST";
  /** 输入 JSON Schema(draft-07);省略 = 任意输入 */
  inputSchema?: object;
}

// ★ 改这里:声明你要暴露的上游 API
const ENDPOINTS: EndpointConfig[] = [
  {
    id: "http.jsonplaceholder.user",
    description: "Fetches a user from JSONPlaceholder by id",
    url: "https://jsonplaceholder.typicode.com/users/{id}",
    method: "GET",
    inputSchema: { type: "object", properties: { id: { type: "integer", minimum: 1 } }, required: ["id"] },
  },
  {
    id: "http.jsonplaceholder.post",
    description: "Creates a post on JSONPlaceholder",
    url: "https://jsonplaceholder.typicode.com/posts",
    method: "POST",
    inputSchema: {
      type: "object",
      properties: { title: { type: "string" }, body: { type: "string" } },
      required: ["title"],
    },
  },
];

/** 把 input 里的 {key} 占位符替换进 URL(仅路径参数) */
function fillUrl(url: string, input: Record<string, unknown>): string {
  return url.replace(/\{(\w+)\}/g, (_, key: string) => encodeURIComponent(String(input[key] ?? "")));
}

export function buildComponents(): ComponentDef[] {
  return ENDPOINTS.map((ep) =>
    defineComponent({
      id: ep.id,
      name: ep.id,
      description: ep.description,
      version: "1.0.0",
      ...(ep.inputSchema !== undefined ? { inputSchema: ep.inputSchema } : {}),
      tags: ["http-proxy"],
      async handle(input) {
        const record = (input ?? {}) as Record<string, unknown>;
        const res = await fetch(fillUrl(ep.url, record), {
          method: ep.method,
          headers: { "content-type": "application/json" },
          ...(ep.method === "POST" ? { body: JSON.stringify(input ?? {}) } : {}),
        });
        const body = (await res.json()) as unknown;
        if (!res.ok) {
          // 上游失败 → ACP 语义(spec §8:50002 UPSTREAM_ERROR)
          throw new AcpError(AcpErrorCode.UPSTREAM_ERROR, `upstream ${res.status}`, body);
        }
        return body;
      },
    })
  );
}

const isMain =
  process.argv[1]?.replace(/\\/g, "/").endsWith("http-proxy.ts") ||
  process.argv[1]?.replace(/\\/g, "/").endsWith("http-proxy.js");
if (isMain) {
  const server = new AcpServer({ name: "http-proxy-node", version: "1.0.0" });
  for (const c of buildComponents()) server.register(c);
  server
    .listen({ port: Number(process.env["ACP_PORT"] ?? 8091) })
    .then(({ port }) => console.log(`HTTP-proxy ACP endpoint: http://localhost:${port}/acp`));
}
